### PR TITLE
Add Frontend Guide

### PR DESCRIPTION
**Pull request by joris974**
This is a WIP. 

I will completed missing sections.
If you think about a missing section please let me know.

### CODE REVIEW

**Comment by jdreaver** [comment]
Awesome!

* Are you going to have an in-depth general JS style guide here, or would that be a separate guide?
* Side question: would you be cool with using Shake to manage the frontend instead of using the scripts directly? That way we won't accidentally diverge between how Shake does things and how your scripts do things. For example, "build all" is just `./build.sh build-frontend`. This includes setup. Also resetting is just `./build.sh clean-frontend` and then a build.

**Comment by eborden** [comment]
Might want to add @stackptr to this review.

File: frontend-guide.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+## Applications
+
+All the frontend codebase exist in [`megarepo/frontend`](https://github.com/frontrowed/megarepo/tree/master/frontend).
+
+The frontend is split in 3 different applications:
+  - [Teacher Dashboard](https://classroom.frontrowed.com)
+  - [Student Dashboard](https://student.frontrowed.com)
+  - [School Dashboard](https://school.frontrowed.com)
+
+All applications use the same stack, share the same tools, build processes, guidelines, architecture, and produce the same type of output.
```
Comment (eborden) [inline frontend-guide.md]: Thank gawd!

File: frontend-guide.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
+# Front Row Frontend
+
+This document is meant to give a broad overview of the state of the frontend codebase of FrontRow.
+
+## Applications
+
+All the frontend codebase exist in [`megarepo/frontend`](https://github.com/frontrowed/megarepo/tree/master/frontend).
+
+The frontend is split in 3 different applications:
+  - [Teacher Dashboard](https://classroom.frontrowed.com)
```
Comment (cdparks) [inline frontend-guide.md]: "split into" or "split between"

File: frontend-guide.md (hunk 1 above)
Comment (cdparks) [inline frontend-guide.md]: "All applications use the same stack; share the same tools, build processes, guidelines, and architecture; and produce the same type of output"

File: frontend-guide.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+| 1024x768        | 26%   |
+| 1280x800        | 9%
+| 1440x900        | 9%    |
+| 1280x1024       | 6%    |
+| 1920x1080       | 4%    |
+
+#### Student Dashboard
+
+For the student dashboard, we optimize for a iPad (1024x768) and ChromeBook (1366x768) sizes.
+All interactions should be thought be used by a touch device.
```
Comment (cdparks) [inline frontend-guide.md]: "All interactions should support using a touch device"

File: frontend-guide.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+Trackpads on ChromeBook is not easy to master and our users have difficulties being precise with it. That's why a lot of ChromeBook users cannot use the drawing features in Math questions.
```
Comment (cdparks) [inline frontend-guide.md]: "Trackpads on Chromebook are not easy to master"

File: frontend-guide.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+| Screen Size     | Users |
+| --------------  |:-----:|
+| 1366x768        | 50%   |
+| 1024x768        | 30%   |
+| 1280x800        | 3%
+| 1440x900        | 3%    |
+
+## Build
+
+All build steps are written in a bash file at the root of each project call `build.sh`.
```
Comment (cdparks) [inline frontend-guide.md]: "called build.sh"

File: frontend-guide.md [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+It contains all the steps to create the different output of each web applications.
```
Comment (cdparks) [inline frontend-guide.md]: "each web application"

File: frontend-guide.md [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+The build usually includes:
+- Converting SASS to CSS
+- Bundling CSS
+- Versionifying CSS bundle
```
Comment (cdparks) [inline frontend-guide.md]: "Versioning"

File: frontend-guide.md [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+
+**Run ESLint with all warnings**
+```
+yarn lint
+```
+
+### Internal dependencies
+
+In order to share code of `frontend-entities` between multiple projects, we import the folder as a package.
+The build process
```
Comment (cdparks) [inline frontend-guide.md]: WIP?

File: frontend-guide.md [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+
+
+### Main tools used for the build
+
+
+#### Browserify
+
+We use Browserify to create JS bundles of our code and vendor files. We use a lib on top of Browserify called [`browserify-incremental`](https://github.com/jsdf/browserify-incremental) allowing us to speed us the bundling process by not re-creating the bundle from scratch after every change.
+Each build type (`dev` or `prod`) use a different cache file for this process (`browserify-cache-development.json`).
```
Comment (cdparks) [inline frontend-guide.md]: "uses a different cache file"

File: frontend-guide.md [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+```
+
+**Launch the watcher and rebuild for dev after each local change**
+```
+yarn watch
+```
+
+**Run ESLint with all warnings**
+```
+yarn lint
```
Comment (eborden) [inline frontend-guide.md]: Might be helpful to add that code climate checks lint errors.

File: frontend-guide.md [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+See Stack/Flow
+
+
+## Stack
+
+The stack of the frontend is almost the same for each project with few exceptions.
+
+### Backbone Collections and Models
+
+When calling an API resource, we use Backbone Collections and Models to transform them into usable, type-checked, javascript usuable class.
```
Comment (cdparks) [inline frontend-guide.md]: "usable", though I would just say objects here probably

File: frontend-guide.md [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+Every resource is atomically defined by it's model. Each model contains a `parse` function validating and transforming data from API response.
```
Comment (cdparks) [inline frontend-guide.md]: I'm not sure what "atomically" means here

File: frontend-guide.md [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+
+_Example of the model parse function_
+```javascript
+parse: function(response: Object): Object {
+  // Validating the API response contains the keys we expect and the values are the type we expect
```
Comment (cdparks) [inline frontend-guide.md]: "Validating the API response contains the keys and values we expect _at runtime_"

File: frontend-guide.md [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+  const newResponse = objectOfType(response, model)
+  // Converting date from string to moment
+  newResponse.createdAt = moment(newResponse.createdAt)
+  newResponse.completedAt = newResponse.completedAt ? moment(newResponse.completedAt) : null
+  return newResponse
+}
+```
+
+To add more type safety around Backbone Model we add some boilerplate around the model creation.
+For each model we define the type of the return of the `get` function for each valid key. Calling `get` on a non valid `key` should throw a flow error, same for using the value as the incorrect type.
```
Comment (cdparks) [inline frontend-guide.md]: "should raise a flow error _at compile time_"

File: frontend-guide.md [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+More info about [Backbone])http://backbonejs.org/). You can also check the [annotated sources of Backbone](http://backbonejs.org/docs/backbone.html)
+
+
+### Backbone Router
+
+TODO
+
+### Flow
+
+[Flow](https://flow.org/) is a tool that add a layer of typing to javascript code. It relies on annotations in the code to figure out which type are the elements. When running `flow check`, flow will check the code base and throw all typing errors found. Flow contains the definition of the javascript API and can be completed by passing definition of libraries we use, like React, Backbone or Lodash. Other library definition can be in [flow-typed](https://github.com/flowtype/flow-typed).
```
Comment (cdparks) [inline frontend-guide.md]: I would say "It relies on annotations in the code to determine the types of expressions". Flow also does some type inference, but it's really weak, and only improves if you have (usually) function-argument annotations.

File: frontend-guide.md [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+Importing an element to a file will also import its type but you can chose to only import its type.
```
Comment (cdparks) [inline frontend-guide.md]: I would say importing "a name into a module imports its type". Or maybe a binding?

File: frontend-guide.md [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+## Testing
+
+We use [mocha](https://mochajs.org/) as the JS test framework and [chai](http://chaijs.com/) as the assertion library.
+
+We expect developers to add tests to logic and helper files. 
+
+All tests exist in the `test` folder of every application. We keep the same name and file path of the tested file for the test file:
+`js/helpers/my-nice-helper.js` should have a corresponding test file: `test/helpers/my-nice-helper-test.js`.
+
+By contention the first `describe` should contain the name of the tested file. Every nested `describe` should contain the name of the tested function.
```
Comment (cdparks) [inline frontend-guide.md]: "By convention"

File: frontend-guide.md [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+Example:
+```javascript
+describe("my-nice-helper", function() {
```
Comment (cdparks) [inline frontend-guide.md]: I would point out why we use `function` instead of arrow functions here. If I recall, `mocha` does something sneaky with `this`

**Review by cdparks — CHANGES_REQUESTED** [review]
Thanks for working on this, Joris.

**Review by eborden — COMMENTED** [review]
Looking great! Thanks for putting this together.

**Review by cdparks — APPROVED** [review]

File: frontend-guide.md (hunk 18 above)
Comment (eborden) [reply frontend-guide.md]: Mmmmm, love me some dynamic binding.

**Pull request merged**